lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -265,6 +273,141 @@ getBlocks logTrace logic recoveryHeadersMessage enqueue nodeId tipHeaderHash che
               Just (MsgBlock block) -> do
                   retrieveBlocksDo conv bvd (i - 1) (block : acc)
 
+-- | Stream some blocks from the network.
+-- Returns Nothing if streaming is disabled by the client or not supported by the peer.
+streamBlocks
+    :: forall t .
+       Trace IO (Severity, Text)
+    -> Maybe DiffusionHealth
+    -> Logic IO
+    -> Word32
+    -> EnqueueMsg
+    -> NodeId
+    -> HeaderHash
+    -> [HeaderHash]
+    -> ((Word32, Maybe Gauge, Conc.TBQueue StreamEntry) -> IO t)
+    -> IO (Maybe t)
+streamBlocks _        _   _     0            _       _      _         _           _ = return Nothing -- Fallback to batch mode
+streamBlocks logTrace smM logic streamWindow enqueue nodeId tipHeader checkpoints k = do
+    blockChan <- atomically $ Conc.newTBQueue $ fromIntegral streamWindow
+    let wqgM = dhStreamWriteQueue <$> smM
+    fallBack <- atomically $ Conc.newTVar False
+    requestVar <- requestBlocks fallBack blockChan
+    r <- k (streamWindow, wqgM, blockChan) `finally` (atomically $ do
+        status <- Conc.readTVar requestVar
+        case status of
+             OQ.PacketAborted -> pure (pure ())
```

Comment:
The outbound queue really does seem to make things difficult here. What to do in case the packet is aborted? We'd want to interrupt the continuation `k`, which could be done by putting `StreamEnd` into the queue, but how to do that?

Reply:
To me it looks like the only way for `requestVar` to become `PacketAborted` is if it `k` has returned and `requestVar` was `PacketEnqueued`.